File: python/src/ambertrace/providers/gemini/interceptor.py
"""Google Gemini SDK interception via monkey-patching.

Supports both Gemini SDKs:
- google-generativeai (original): patches GenerativeModel.generate_content
- google-genai (newer): patches the models resource generate_content
"""

import functools
import logging
import time
import uuid
from typing import Any, Callable, Optional

from ambertrace.providers.base import BaseInterceptor
from ambertrace.providers.gemini.collector import GeminiCollector
from ambertrace.transport import get_transport

logger = logging.getLogger(__name__)


class GeminiInterceptor(BaseInterceptor):
    """Handles monkey-patching of Google Gemini SDK methods.

    Supports both the original google-generativeai SDK and the newer
    google-genai SDK. Each is patched independently.
    """

    def __init__(self) -> None:
        """Initialize interceptor."""
        # Original SDK (google-generativeai)
        self._original_genai_sync: Optional[Callable] = None
        self._original_genai_async: Optional[Callable] = None
        self._genai_patched = False

        # Newer SDK (google-genai)
        self._original_genai2_sync: Optional[Callable] = None
        self._original_genai2_async: Optional[Callable] = None
        self._genai2_patched = False

        self._collector = GeminiCollector()

    def get_provider_name(self) -> str:
        """Return provider name."""
        return "gemini"

    def patch(self) -> None:
        """Apply monkey-patches to Gemini SDK methods.

        Attempts to patch both Gemini SDKs independently. If only one
        is installed, only that one is patched.
        """
        if self._genai_patched and self._genai2_patched:
            logger.debug("Gemini already fully patched, skipping")
            return

        # --- Original SDK: google.generativeai ---
        if not self._genai_patched:
            try:
                import google.generativeai as genai

                model_class = genai.GenerativeModel

                # Patch sync: GenerativeModel.generate_content
                if hasattr(model_class, "generate_content"):
                    self._original_genai_sync = model_class.generate_content
                    model_class.generate_content = self._wrap_genai_sync(
                        self._original_genai_sync
                    )
                    logger.debug("Patched GenerativeModel.generate_content (sync)")

                # Patch async: GenerativeModel.generate_content_async
                if hasattr(model_class, "generate_content_async"):
                    self._original_genai_async = model_class.generate_content_async
                    model_class.generate_content_async = self._wrap_genai_async(
                        self._original_genai_async
                    )
                    logger.debug("Patched GenerativeModel.generate_content_async (async)")

                self._genai_patched = True
                logger.info("google-generativeai SDK patching completed")

            except ImportError:
                logger.debug("google-generativeai SDK not found, skipping")
            except Exception as e:
                logger.error(f"Failed to patch google-generativeai: {e}", exc_info=True)

        # --- Newer SDK: google.genai ---
        if not self._genai2_patched:
            try:
                from google import genai  # noqa: F811

                # The newer SDK uses client.models.generate_content()
                # We need to find the Models resource class
                models_module = genai.models if hasattr(genai, "models") else None

                if models_module is not None:
                    # Try to find the Models class
                    models_class = None
                    if hasattr(models_module, "Models"):
                        models_class = models_module.Models
                    elif hasattr(models_module, "AsyncModels"):
                        # Try to get the sync class from the module
                        for attr_name in dir(models_module):
                            attr = getattr(models_module, attr_name)
                            if (
                                isinstance(attr, type)
                                and hasattr(attr, "generate_content")
                                and attr_name != "AsyncModels"
                            ):
                                models_class = attr
                                break

                    if models_class and hasattr(models_class, "generate_content"):
                        self._original_genai2_sync = models_class.generate_content
                        models_class.generate_content = self._wrap_genai2_sync(
                            self._original_genai2_sync
                        )
                        logger.debug("Patched google.genai Models.generate_content (sync)")

                    # Try to patch async variant
                    async_models_class = None
                    if hasattr(models_module, "AsyncModels"):
                        async_models_class = models_module.AsyncModels

                    if async_models_class and hasattr(async_models_class, "generate_content"):
                        self._original_genai2_async = async_models_class.generate_content
                        async_models_class.generate_content = self._wrap_genai2_async(
                            self._original_genai2_async
                        )
                        logger.debug("Patched google.genai AsyncModels.generate_content (async)")

                    self._genai2_patched = True
                    logger.info("google-genai SDK patching completed")
                else:
                    logger.warning("Could not find models module in google.genai")

            except ImportError:
                logger.debug("google-genai SDK not found, skipping")
            except Exception as e:
                logger.error(f"Failed to patch google-genai: {e}", exc_info=True)

        if self.is_patched():
            logger.info("Gemini SDK patching completed")

    def unpatch(self) -> None:
        """Remove monkey-patches and restore original Gemini methods."""
        # Restore original SDK (google-generativeai)
        if self._genai_patched:
            try:
                import google.generativeai as genai

                model_class = genai.GenerativeModel

                if self._original_genai_sync is not None:
                    model_class.generate_content = self._original_genai_sync
                    logger.debug("Restored GenerativeModel.generate_content (sync)")

                if self._original_genai_async is not None:
                    model_class.generate_content_async = self._original_genai_async
                    logger.debug("Restored GenerativeModel.generate_content_async (async)")

                self._genai_patched = False
                self._original_genai_sync = None
                self._original_genai_async = None
                logger.info("google-generativeai SDK unpatching completed")

            except Exception as e:
                logger.error(
                    f"Failed to unpatch google-generativeai: {e}", exc_info=True
                )

        # Restore newer SDK (google-genai)
        if self._genai2_patched:
            try:
                from google import genai

                models_module = genai.models if hasattr(genai, "models") else None

                if models_module is not None:
                    if self._original_genai2_sync is not None:
                        models_class = None
                        if hasattr(models_module, "Models"):
                            models_class = models_module.Models
                        if models_class:
                            models_class.generate_content = self._original_genai2_sync
                            logger.debug("Restored google.genai Models.generate_content")

                    if self._original_genai2_async is not None:
                        async_models_class = None
                        if hasattr(models_module, "AsyncModels"):
                            async_models_class = models_module.AsyncModels
                        if async_models_class:
                            async_models_class.generate_content = self._original_genai2_async
                            logger.debug("Restored google.genai AsyncModels.generate_content")

                self._genai2_patched = False
                self._original_genai2_sync = None
                self._original_genai2_async = None
                logger.info("google-genai SDK unpatching completed")

            except Exception as e:
                logger.error(f"Failed to unpatch google-genai: {e}", exc_info=True)

    def is_patched(self) -> bool:
        """Check if patches are currently applied."""
        return self._genai_patched or self._genai2_patched

    # --- Wrappers for original SDK (google-generativeai) ---

    def _wrap_genai_sync(self, original_method: Callable) -> Callable:
        """Create wrapper for sync GenerativeModel.generate_content.

        The original SDK stores the model name on the instance, not in
        method arguments. The wrapper extracts it and injects it as
        _ambertrace_model for the collector.
        """

        @functools.wraps(original_method)
        def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            # Extract model name from the GenerativeModel instance
            model_name = getattr(self_instance, "model_name", None) or getattr(
                self_instance, "_model_name", "unknown"
            )

            # Build enriched kwargs for the collector
            enriched_kwargs = {**kwargs, "_ambertrace_model": model_name}
            # Capture contents from positional args if not in kwargs
            if args and "contents" not in enriched_kwargs:
                enriched_kwargs["contents"] = args[0]

            try:
                response = original_method(self_instance, *args, **kwargs)

                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=enriched_kwargs,
                        response=response,
                        error=None,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as e:
                    logger.error(
                        f"Error collecting Gemini trace {trace_id}: {e}", exc_info=True
                    )

                return response

            except Exception as e:
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=enriched_kwargs,
                        response=None,
                        error=e,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Gemini error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )
                raise

        return wrapper

    def _wrap_genai_async(self, original_method: Callable) -> Callable:
        """Create wrapper for async GenerativeModel.generate_content_async."""

        @functools.wraps(original_method)
        async def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            model_name = getattr(self_instance, "model_name", None) or getattr(
                self_instance, "_model_name", "unknown"
            )
            enriched_kwargs = {**kwargs, "_ambertrace_model": model_name}
            if args and "contents" not in enriched_kwargs:
                enriched_kwargs["contents"] = args[0]

            try:
                response = await original_method(self_instance, *args, **kwargs)

                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=enriched_kwargs,
                        response=response,
                        error=None,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace_async(trace_dict)
                except Exception as e:
                    logger.error(
                        f"Error collecting Gemini trace {trace_id}: {e}", exc_info=True
                    )

                return response

            except Exception as e:
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=enriched_kwargs,
                        response=None,
                        error=e,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace_async(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Gemini error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )
                raise

        return wrapper

    # --- Wrappers for newer SDK (google-genai) ---

    def _wrap_genai2_sync(self, original_method: Callable) -> Callable:
        """Create wrapper for google.genai Models.generate_content.

        The newer SDK passes model as a kwarg:
            client.models.generate_content(model="gemini-2.0-flash", contents="Hello")
        """

        @functools.wraps(original_method)
        def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            try:
                response = original_method(self_instance, *args, **kwargs)

                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=response,
                        error=None,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as e:
                    logger.error(
                        f"Error collecting Gemini trace {trace_id}: {e}", exc_info=True
                    )

                return response

            except Exception as e:
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=None,
                        error=e,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Gemini error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )
                raise

        return wrapper

    def _wrap_genai2_async(self, original_method: Callable) -> Callable:
        """Create wrapper for google.genai AsyncModels.generate_content."""

        @functools.wraps(original_method)
        async def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            try:
                response = await original_method(self_instance, *args, **kwargs)

                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=response,
                        error=None,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace_async(trace_dict)
                except Exception as e:
                    logger.error(
                        f"Error collecting Gemini trace {trace_id}: {e}", exc_info=True
                    )

                return response

            except Exception as e:
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=None,
                        error=e,
                    )
                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace_async(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Gemini error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )
                raise

        return wrapper

File: typescript/src/providers/gemini/interceptor.ts
/**
 * Google Gemini SDK interceptor using prototype manipulation.
 *
 * Supports both Gemini SDKs:
 * - @google/generative-ai (original SDK): patches GenerativeModel.prototype.generateContent
 * - @google/genai (newer SDK): patches models resource class generateContent
 *
 * Intercepts generateContent() calls to collect traces transparently.
 */

import { randomUUID } from 'crypto';
import type { BaseInterceptor } from '../base';
import { GeminiCollector } from './collector';
import { getTransport } from '../../transport';
import { getConfig } from '../../config';

/**
 * Intercepts Google Gemini SDK content generation calls.
 *
 * Supports patching both the original @google/generative-ai SDK and
 * the newer @google/genai SDK independently.
 */
export class GeminiInterceptor implements BaseInterceptor {
  private collector = new GeminiCollector();

  // Original SDK state
  private genaiPatched = false;
  private originalGenaiGenerateContent: ((...args: unknown[]) => unknown) | null = null;
  private genaiPrototype: Record<string, unknown> | null = null;

  // Newer SDK state
  private genai2Patched = false;
  private originalGenai2GenerateContent: ((...args: unknown[]) => unknown) | null = null;
  private genai2Prototype: Record<string, unknown> | null = null;

  getProviderName(): string {
    return 'google';
  }

  patch(): void {
    // Patch original SDK
    if (!this.genaiPatched) {
      try {
        this.patchOriginalSdk();
      } catch {
        // Original SDK not available, skip
      }
    }

    // Patch newer SDK
    if (!this.genai2Patched) {
      try {
        this.patchNewerSdk();
      } catch {
        // Newer SDK not available, skip
      }
    }

    if (getConfig()?.debug) {
      const parts: string[] = [];
      if (this.genaiPatched) parts.push('@google/generative-ai');
      if (this.genai2Patched) parts.push('@google/genai');
      if (parts.length > 0) {
        console.log(`[AmberTrace] Gemini SDK patched: ${parts.join(', ')}`);
      }
    }
  }

  unpatch(): void {
    // Unpatch original SDK
    if (this.genaiPatched && this.genaiPrototype && this.originalGenaiGenerateContent) {
      this.genaiPrototype.generateContent = this.originalGenaiGenerateContent;
      this.genaiPatched = false;
      this.originalGenaiGenerateContent = null;
      this.genaiPrototype = null;
    }

    // Unpatch newer SDK
    if (this.genai2Patched && this.genai2Prototype && this.originalGenai2GenerateContent) {
      this.genai2Prototype.generateContent = this.originalGenai2GenerateContent;
      this.genai2Patched = false;
      this.originalGenai2GenerateContent = null;
      this.genai2Prototype = null;
    }

    if (getConfig()?.debug) {
      console.log('[AmberTrace] Gemini SDK unpatched');
    }
  }

  isPatched(): boolean {
    return this.genaiPatched || this.genai2Patched;
  }

  private patchOriginalSdk(): void {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const genai = require('@google/generative-ai');

    // Try to find GenerativeModel class
    const ModelClass = genai.GenerativeModel;
    if (!ModelClass?.prototype) {
      throw new Error('GenerativeModel not found in @google/generative-ai');
    }

    const prototype = ModelClass.prototype;

    if (typeof prototype.generateContent !== 'function') {
      throw new Error('generateContent is not a function');
    }

    // Store original
    this.originalGenaiGenerateContent = prototype.generateContent;
    this.genaiPrototype = prototype;

    // Create wrapper - capture reference as non-null (we verified above)
    const originalMethod = this.originalGenaiGenerateContent!;
    const self = this;

    prototype.generateContent = function wrappedGenerateContent(
      this: Record<string, unknown>,
      ...args: unknown[]
    ) {
      const config = getConfig();

      // If tracing is disabled, call original method
      if (!config?.enabled) {
        return originalMethod.apply(this, args);
      }

      // Extract model name from instance
      const modelName = this.model ?? 'unknown';

      // Build request kwargs
      const requestArgs: Record<string, unknown> = {
        _ambertrace_model: modelName,
      };

      // Extract contents from positional args
      if (args.length > 0) {
        requestArgs.contents = args[0];
      }

      // Merge any remaining kwargs
      if (args.length > 1 && typeof args[1] === 'object' && args[1] !== null) {
        Object.assign(requestArgs, args[1] as Record<string, unknown>);
      }

      // Generate trace ID and start time
      const traceId = randomUUID();
      const startTime = Date.now();

      try {
        const result = originalMethod.apply(this, args);

        // Check if result is a promise
        if (result && typeof (result as Promise<unknown>).then === 'function') {
          return (result as Promise<unknown>)
            .then((response) => {
              self.collectAndSendTrace(traceId, startTime, requestArgs, response);
              return response;
            })
            .catch((error: Error) => {
              self.collectAndSendTrace(traceId, startTime, requestArgs, undefined, error);
              throw error;
            });
        }

        // Synchronous result
        self.collectAndSendTrace(traceId, startTime, requestArgs, result);
        return result;
      } catch (error) {
        self.collectAndSendTrace(traceId, startTime, requestArgs, undefined, error as Error);
        throw error;
      }
    };

    this.genaiPatched = true;
  }

  private patchNewerSdk(): void {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const genai = require('@google/genai');

    // Newer SDK exposes models resource, find the prototype to patch
    const GoogleGenAI = genai.GoogleGenAI;
    if (!GoogleGenAI) {
      throw new Error('GoogleGenAI not found in @google/genai');
    }

    // The newer SDK has a Models class that has generateContent
    // We need to find it - it could be on the GoogleGenAI instance's models property
    const ModelsClass = genai.Models ?? genai.GoogleGenAI?.Models;
    if (!ModelsClass?.prototype) {
      throw new Error('Models class not found in @google/genai');
    }

    const prototype = ModelsClass.prototype;

    if (typeof prototype.generateContent !== 'function') {
      throw new Error('generateContent is not a function on Models');
    }

    // Store original
    this.originalGenai2GenerateContent = prototype.generateContent;
    this.genai2Prototype = prototype;

    // Create wrapper - capture reference as non-null (we verified above)
    const originalMethod = this.originalGenai2GenerateContent!;
    const self = this;

    prototype.generateContent = function wrappedGenerateContent(
      this: Record<string, unknown>,
      ...args: unknown[]
    ) {
      const config = getConfig();

      // If tracing is disabled, call original method
      if (!config?.enabled) {
        return originalMethod.apply(this, args);
      }

      // Newer SDK: generateContent({ model, contents, ... })
      const requestArgs: Record<string, unknown> = {};
      if (args.length > 0 && typeof args[0] === 'object' && args[0] !== null) {
        Object.assign(requestArgs, args[0] as Record<string, unknown>);
      }

      // Generate trace ID and start time
      const traceId = randomUUID();
      const startTime = Date.now();

      try {
        const result = originalMethod.apply(this, args);

        // Check if result is a promise
        if (result && typeof (result as Promise<unknown>).then === 'function') {
          return (result as Promise<unknown>)
            .then((response) => {
              self.collectAndSendTrace(traceId, startTime, requestArgs, response);
              return response;
            })
            .catch((error: Error) => {
              self.collectAndSendTrace(traceId, startTime, requestArgs, undefined, error);
              throw error;
            });
        }

        // Synchronous result
        self.collectAndSendTrace(traceId, startTime, requestArgs, result);
        return result;
      } catch (error) {
        self.collectAndSendTrace(traceId, startTime, requestArgs, undefined, error as Error);
        throw error;
      }
    };

    this.genai2Patched = true;
  }

  private collectAndSendTrace(
    traceId: string,
    startTime: number,
    requestArgs: Record<string, unknown>,
    response?: unknown,
    error?: Error
  ): void {
    try {
      // Collect trace
      const trace = this.collector.collectTrace(traceId, startTime, requestArgs, response, error);

      if (trace) {
        // Send trace asynchronously
        const transport = getTransport();
        transport.sendTrace(trace);
      }
    } catch (err) {
      // Never throw from trace collection
      if (getConfig()?.debug) {
        console.error('[AmberTrace] Error collecting/sending Gemini trace:', err);
      }
    }
  }
}
